packages/token-generation-readmodel-checker-verifier/src/utils/utils.ts

```diff
@@ -0,0 +1,1254 @@
+        ? [
+            extractIdsFromGSIPKConsumerIdEServiceId(
+              d.GSIPK_consumerId_eserviceId
+            )?.eserviceId,
+          ]
+        : []
+    ),
+    ...dataC.map((d) => getIdentificationKey(d)),
+  ]);
+  return Array.from(allIds).map((id) => [
+    dataA.find(
+      (d: PlatformStatesCatalogEntry) => getIdentificationKey(d) === id
+    ),
+    dataB.filter(
+      (d: TokenGenerationStatesClientPurposeEntry) =>
+        extractIdsFromGSIPKEServiceIdDescriptorId(
+          d.GSIPK_eserviceId_descriptorId
+        )?.eserviceId === id ||
+        extractIdsFromGSIPKConsumerIdEServiceId(d.GSIPK_consumerId_eserviceId)
+          ?.eserviceId === id
+    ),
+    dataC.find((d: EService) => getIdentificationKey(d) === id),
+  ]);
+}
+
+export function countCatalogDifferences(
+  differences: CatalogDifferencesResult,
+  logger: Logger
+): number {
+  // eslint-disable-next-line functional/no-let
+  let differencesCount = 0;
+  differences.forEach(([platformCatalog, tokenCatalog, readModelEService]) => {
+    if (!readModelEService) {
+      if (platformCatalog) {
+        logger.error(
+          `Read model eservice not found for ${
+            extractIdFromPlatformStatesPK(platformCatalog.PK).id
+          }`
+        );
+      } else if (tokenCatalog?.[0].GSIPK_eserviceId_descriptorId) {
+        logger.error(
+          `Read model eservice not found for ${
+            extractIdsFromGSIPKEServiceIdDescriptorId(
+              tokenCatalog[0].GSIPK_eserviceId_descriptorId
+            )?.eserviceId
+          }`
+        );
+      } else {
+        throw genericInternalError(
+          "Unexpected error while counting catalog differences"
+        );
+      }
+      differencesCount++;
+    } else if (readModelEService) {
+      logger.error(
+        `Catalog states are not equal:
+  platform-states entry: ${JSON.stringify(platformCatalog)}
+  token-generation-states entries: ${JSON.stringify(tokenCatalog)}
+  purpose read-model: ${JSON.stringify(readModelEService)}`
+      );
+      differencesCount++;
+    }
+  });
+
+  return differencesCount;
+}
+
+// clients
+export async function compareReadModelClientsWithTokenGenReadModel({
+  platformStatesEntries,
+  tokenGenerationStatesEntries,
+  readModel,
+}: {
+  platformStatesEntries: PlatformStatesClientEntry[];
+  tokenGenerationStatesEntries: TokenGenerationStatesGenericEntry[];
+  readModel: ReadModelRepository;
+}): Promise<ClientDifferencesResult> {
+  const readModelService = readModelServiceBuilder(readModel);
+  const [resultsA, resultsB, resultsC] = await Promise.all([
+    platformStatesEntries,
+    tokenGenerationStatesEntries,
+    readModelService.getAllReadModelClients(),
+  ]);
+  return zipClientDataById(
+    resultsA,
+    resultsB,
+    resultsC
+  ).reduce<ClientDifferencesResult>((acc, [a, b, c]) => {
+    // TODO: are missing token entries considered errors or not?
+    if (!c || (c && (!a || b?.length === 0))) {
```

Comment:
Are missing token-generation-states entries considered errors or not?

Reply:
yes, they are if the rows should exist.
For each client, we expect
- if the client does not contain keys, the number of records in token-generation-states should be zero
- if the client does not contain purposes, the number of records in token-generation-states should be equal to the number of keys in the client
- if the client contains purposes, the number of records in token-generation-states should be equal to the number of keys in the client multiplied by the number of purposes